src/sqlfluff/rules/L042.py

```diff
@@ -48,44 +70,281 @@ class Rule_L042(BaseRule):
+                ),
+                WhitespaceSegment(),
+                segmentify("AS"),
+                WhitespaceSegment(),
+                subquery,
+            )
+        )
+        # (mypy is not understanding the generic)
+        ctes.insert_cte(new_cte)  # type: ignore
+        # TODO: Create non-mutative helper function.
+        # We will replace the whole tree, mutate the table expression.
+        this_seg[0].segments = (
+            TableExpressionSeg(
+                segments=(
+                    TableReferenceSeg(
+                        segments=(
+                            CodeSegment(
+                                raw=alias_name,
+                                name="naked_identifier",
+                                type="identifier",
+                            ),
+                        )
+                    ),
                 )
-                if seg:
-                    return LintResult(
-                        anchor=seg[0],
-                        description=f"{parent_type} clauses should not contain "
-                        "subqueries. Use CTEs instead",
-                    )
-        return None
+            ),
+        )
+        res = LintResult(
+            anchor=subquery,
+            description=f"{parent_type} clauses should not contain "
+            "subqueries. Use CTEs instead",
+            fixes=[],
+        )
+        lint_results.append(res)
+
+    if ctes.has_duplicates() or is_recursive:
+        # If we have duplicate CTE names just don't fix anything
+        return lint_results
+
+    new_select = WithCompoundStatementSegment(
+        segments=tuple(
+            [
+                segmentify("WITH"),
+                WhitespaceSegment(),
+                *ctes.get_segements(),
+                NewlineSegment(),
+                output_select[0],
+            ]
+        )
+    )
+    # Add fixes to the last result only
+    res = lint_results.pop()
+    res.fixes = [LintFix.replace(root_select[0], edit_segments=[new_select])]
+    lint_results.append(res)
+    return lint_results
+
+
+def _is_child(maybe_parent: Segments, maybe_child: Segments) -> bool:
+    """Is the child actually between the start and end markers of the parent."""
+    assert len(maybe_child) == 1, "Cannot assess Childness of multiple Segments"
+
+    child_markers = maybe_child[0].pos_marker
+    if not child_markers:
+        return False  # pragma: no cover
+    for segement in maybe_parent:
+        parent_pos = segement.pos_marker
+        if not parent_pos:
+            continue  # pragma: no cover
+
+        if child_markers < parent_pos.start_point_marker():
+            continue  # pragma: no cover
+
+        if child_markers > parent_pos.end_point_marker():
+            continue
+
+        return True
+
+    return False
+
+
+def _find_nested_subqueries(
+    select: Segments,
+) -> Generator[Tuple[str, Segments, Segments, BaseSegment], None, None]:
```

Comment:
Think we could create a `NamedTuple` declaration for this? Could be a bit more readable.